ksqldb-engine/src/main/java/io/confluent/ksql/utilization/PersistentQuerySaturationMetrics.java

```diff
@@ -0,0 +1,365 @@
+        perThreadSaturation.remove(threadName);
+        reporter.cleanup(QUERY_THREAD_SATURATION, ImmutableMap.of(THREAD_ID, threadName));
+      }
+      return saturation;
+    }
+
+    private void cleanup(final MetricsReporter reporter) {
+      for (final String threadName : perThreadSaturation.keySet()) {
+        reporter.cleanup(QUERY_THREAD_SATURATION, ImmutableMap.of(THREAD_ID, threadName));
+      }
+      for (final QueryId queryId : queryIds) {
+        reporter.cleanup(QUERY_SATURATION, ImmutableMap.of(QUERY_ID, queryId.toString()));
+      }
+    }
+  }
+
+  private static final class ThreadSaturation {
+    private final String threadName;
+    private final List<BlockedTimeSample> samples = new LinkedList<>();
+    private final Instant startTime;
+    private final Duration window;
+    private final Duration sampleMargin;
+
+    private ThreadSaturation(
+        final String threadName,
+        final long startTime,
+        final Duration window,
+        final Duration sampleMargin
+    ) {
+      this.threadName = Objects.requireNonNull(threadName, "threadName");
+      this.startTime = Instant.ofEpochMilli(startTime);
+      this.window = Objects.requireNonNull(window, "window");
+      this.sampleMargin = Objects.requireNonNull(sampleMargin, "sampleMargin");
+    }
+
+    private boolean inRange(final Instant time, final Instant start, final Instant end) {
+      return time.isAfter(start) && time.isBefore(end);
+    }
+
+    private Optional<Double> measure(final Instant now, final BlockedTimeSample current) {
+      final Instant windowStart = now.minus(window);
+      final Instant earliest = now.minus(window.plus(sampleMargin));
+      final Instant latest = now.minus(window.minus(sampleMargin));
+      LOGGER.debug(
+          "{}: record and measure with now {},  window {} ({} : {})",
+          threadName,
+          now,
+          windowStart,
+          earliest,
+          latest
+      );
+      samples.add(current);
+      samples.removeIf(s -> s.timestamp.isBefore(earliest));
+      if (!inRange(samples.get(0).timestamp, earliest, latest) && !startTime.isAfter(windowStart)) {
+        return Optional.empty();
+      }
+      final BlockedTimeSample startSample = samples.get(0);
+      LOGGER.debug("{}: start sample {}", threadName, startSample);
+      double blocked =
+          Math.max(current.totalBlockedTime - startSample.totalBlockedTime, 0);
+      Instant observedStart = samples.get(0).timestamp;
+      if (startTime.isAfter(windowStart)) {
+        LOGGER.debug("{}: start time {} is after window start", threadName, startTime);
+        blocked += Duration.between(windowStart, startTime).toNanos();
+        observedStart = windowStart;
+      }
+
+      final Duration duration = Duration.between(observedStart, current.timestamp);
+      final double durationNs = duration.toNanos();
+      return Optional.of((durationNs - Math.min(blocked, durationNs)) / durationNs);
```

Comment:
@rodesai - Why did we take the difference between the time interval and blocked time to compute the saturation. Shouldn't this be just `Math.min(blocked, durationNs) / duration`